### DISCUSSION TITLE
"AttributeError: __pydantic_model__" when response_model is union

### DISCUSSION THREAD

**Discussion by chbndrhnns** [fastapi/fastapi#6300] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typing

import fastapi
import httpx
import pydantic as pydantic
import pytest
from fastapi import FastAPI
from fastapi.encoders import jsonable_encoder
from pydantic import ValidationError
from starlette import status
from starlette.responses import JSONResponse


class ResponseOne(pydantic.BaseModel):
    val: str


class ResponseTwo(pydantic.BaseModel):
    val: str


app = FastAPI()
incorrect_response = {"abc": "def"}


@app.exception_handler(ValidationError)
async def validation_exception_handler(request: fastapi.Request, exc: ValidationError):
    return JSONResponse(
        status_code=status.HTTP_400_BAD_REQUEST,
        content=jsonable_encoder({"detail": exc.errors()}),
    )


@app.get("/union", response_model=typing.Union[ResponseOne, ResponseTwo])
async def union():
    return incorrect_response


@app.get("/single", response_model=ResponseOne)
async def single():
    return incorrect_response


@pytest.fixture
async def client():
    async with httpx.AsyncClient(app=app, base_url="http://localhost") as client:
        yield client


async def test_unprintable(client):
    res = await client.get("/union")
    assert res.status_code == 400


async def test_400(client):
    res = await client.get("/single")
    assert res.status_code == 400
```

### Description

When `response_model` is set to a union of two possible models and I return something that does not match one of the two models, FastAPI returns a 500 error:

```
test_.py:30: in validation_exception_handler
    content=jsonable_encoder({"detail": exc.errors()}),
../../../../../Library/Caches/pypoetry/virtualenvs/bla/lib/python3.9/site-packages/pydantic/error_wrappers.py:63: in errors
    config = self.model.__pydantic_model__.__config__  # type: ignore
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = typing.Union[tests.test_.ResponseOne, tests.test_.ResponseTwo]
attr = '__pydantic_model__'

    def __getattr__(self, attr):
        # We are careful for copy and pickle.
        # Also for simplicity we just don't relay all dunder names
        if '__origin__' in self.__dict__ and not _is_dunder(attr):
            return getattr(self.__origin__, attr)
>       raise AttributeError(attr)
E       AttributeError: __pydantic_model__

../../../../../.pyenv/versions/3.9.9/lib/python3.9/typing.py:710: AttributeError
```

### Operating System

macOS

### FastAPI Version

0.81.0

### Python Version

3.9.9

### Additional Context

It also does not work in FastAPI 0.7* and the pydantic version (1.9 vs 1.0) has no influence, either.

**Comment by chbndrhnns** [comment]
Here is a test with pure pydantic which shows the behaviour.

`ValidationError` seems not to support a union of models for the `model` parameter.

```py
def test_error_from_union():
    class One(BaseModel):
        val: int

    class Two(BaseModel):
        val: int

    with pytest.raises(ValidationError) as err_one:
        One()

    with pytest.raises(ValidationError) as err_two:
        Two(val="abc")

    model = typing.Union[One, Two]

    error = ValidationError([err_one, err_two], model=model)
    assert error.errors()

```

**Comment by mbroton** [comment]
It looks like a bug.
Validation is just fine, but if any validation errors occured, this line (137) in `routing.py` will be executed:
```py3
if errors:
        raise ValidationError(errors, field.type_)
```
and at that time, `field.type_` is exactly `typing.Union[One, Two]`
This is then passed to Pydantic's `error_wrappers.py`:
```py3
class ValidationError(Representation, ValueError):
    __slots__ = 'raw_errors', 'model', '_error_cache'

    def __init__(self, errors: Sequence[ErrorList], model: 'ModelOrDc') -> None:
        self.raw_errors = errors
        self.model = model
        self._error_cache: Optional[List['ErrorDict']] = None

    def errors(self) -> List['ErrorDict']:
        if self._error_cache is None:
            try:
                config = self.model.__config__  # type: ignore
            except AttributeError:
                config = self.model.__pydantic_model__.__config__  # type: ignore
            self._error_cache = list(flatten_errors(self.raw_errors, config))
        return self._error_cache
```
and we have AttributeError because `typing.Union` don't have `__config__` or `__pydantic_model__` attributes.

**Comment by iudeen** [comment]
So it's more of Pydantic issue you mean to say?

**Comment by mbroton** [comment]
Hard to say, but I would say it's FastAPI's problem.
I'm still new to FastAPI, so I don't understand it well.

Here's why I think it's a problem with FastAPI
In this case:
```py3
class One(BaseModel):
    value: int


class Two(BaseModel):
    val: int


@app.get("/", response_model=typing.Union[One, Two])
def home():
    return {"valuee": 1}  # wrong
```
response is properly validated (proper errors are generated):
![image](https://user-images.githubusercontent.com/50829834/188955726-ed9501b6-ab75-4a3f-abf8-abd1c3cb10ff.png)
but `field.type_` doesn't have `__config__` (nor `__pydantic_model__`) attribute and this is the cause of the problem (Pydantic is expecting it)
![image](https://user-images.githubusercontent.com/50829834/188957605-931fe4a8-20f8-4184-b20e-5721e0781e2c.png)
on the screenshot you can see that the `type_.__config__` attribute is missing.

On the other hand, in this case (I would say it's similar, because response_model is an object from `typing`, but the difference is that it takes single argument):
```py3
class One(BaseModel):
    value: int


class Two(BaseModel):
    val: int


@app.get("/", response_model=typing.List[One])
def home():
    return {"valuee": 1}  # also wrong, but response_model is typing.List
```
response is also properly validated (proper errors), but `field.type_` contains `__config__` attribute:
![image](https://user-images.githubusercontent.com/50829834/188958034-d15d7bdf-a3b6-4d74-9641-02f340b757be.png)
and there's no exception from Pydantic - it's working well.

These fields are created by `create_cloned_field` so I would be looking there for an answer.
Another difference that I noticed that feels like potential clue is that in case of `typing.List`, `type_` is a class variable and its type is `pydantic.main.One`, but in a case of `typing.Union`, `type_` is an attribute of an instance (if the debugger is not tricking me) and its type is `typing.Union[One, Two]`.

**TLDR:**
If `response_model=typing.List[SomeModel]`, the `ModelField` in `serialize_response` has attribute `type_.__config__`, but if `response_model=typing.Union[SomeModel1, SomeModel2]`, it hasn't.
In case of errors, `field.type_` is passed to `pydantic.ValidationError` which expects `field.type_` to have `__config__` attribute.
It hasn't, so it raises `AttributeError`.

**Comment by JarroVGIT** [comment]
Is this a duplicate of https://github.com/tiangolo/fastapi/issues/3772 ?

**Comment by chbndrhnns** [comment]
@JarroVGIT thanks for bringing it up. Yes, it is.
